zebrad/src/commands/start/sync.rs

```diff
@@ -198,100 +192,71 @@ where
+            download_set
+                .into_iter()
+                .chain(self.prospective_tips.iter().cloned())
+                .collect(),
+        )
+        .await?;
 
         Ok(())
     }
 
     /// Queue downloads for each block that isn't currently known to our node
-    async fn request_blocks(&mut self, mut hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
-        hashes.retain(|hash| !self.known_block(hash));
-
+    async fn request_blocks(&mut self, hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
         for chunk in hashes.chunks(10usize) {
-            self.queue_download(chunk).await?;
-        }
-
-        Ok(())
-    }
-
-    /// Drive block downloading futures to completion and dispatch downloaded
-    /// blocks to the validator
-    async fn process_blocks(&mut self) -> Result<(), Report> {
-        info!(in_flight = self.block_requests.len(), "processing blocks");
-
-        while let Some(res) = self.block_requests.next().await {
-            match res.map_err::<Report, _>(|e| eyre!(e)) {
-                Ok(zn::Response::Blocks(blocks)) => {
-                    info!(count = blocks.len(), "received blocks");
-                    for block in blocks {
-                        let hash = block.as_ref().into();
-                        assert!(
-                            self.downloading.remove(&hash),
-                            "all received blocks should be explicitly requested and received once"
-                        );
-                        let _ = self.downloaded.insert(hash);
-                        self.validate_block(block).await?;
+            let set = chunk.iter().cloned().collect();
+
+            let request = self
+                .peer_set
+                .ready_and()
+                .await
+                .map_err(|e| eyre!(e))?
+                .call(zn::Request::BlocksByHash(set));
+
+            let mut state = self.state.clone();
+
+            let _ = tokio::spawn(async move {
+                match async move {
+                    let resp = request.await?;
+
+                    if let zn::Response::Blocks(blocks) = resp {
+                        debug!(count = blocks.len(), "received blocks");
+
+                        for block in blocks {
+                            state
+                                .ready_and()
+                                .await?
+                                .call(zs::Request::AddBlock { block })
+                                .await?;
+                        }
+                    } else {
+                        debug!(?resp, "unexpected response");
                     }
+
+                    Ok::<_, Error>(())
                 }
-                Ok(_) => continue,
-                Err(e) => {
-                    error!("{:?}", e);
+                .await
+                {
+                    Ok(()) => {}
+                    Err(e) => {
+                        // TODO(jlusby): retry policy?
```

Comment:
Instead of dealing with retries here, the requests should be made against a retry layer, so that an error in block downloading indicates a (more) permanent failure requiring user intervention rather than an ephemeral one.